### PR TITLE
feat: decodeUriParameters option

### LINKED ISSUE #208: fast-decode-uri-component isn't really fast
### fast-decode-uri-component isn't really fast

During my investigation of #206 problem I was playing with `fast-decode-uri-component` in order to create `fast-decode-uri` that would work exactly as `decodeURI`. During my own benchmarking I was never able to beat the standard `decodeURI`. The reason for that, as I found later, was the test data. 

In `fast-decode-uri-component` they use a lot of malformed URIs for benchmarking. The standard `decodeURIComponent` processes malformed URIs extremely slow. But when you use proper URIs the default function works 2-3 times faster than the "fast" one. 

How many URIs are malformed on a highly loaded API? The only way to get malformed URI is to manually mess it up in browser or misconfigure a new integration (which eventually will be fixed). It's quite counterproductive to optimize the framework to process malformed URIs faster and process proper URIs slower.

My suggestion is to get rid of the `fast-decode-uri-component` and use safe and a little fastified version of the standard function like I show in benchmark below. That will work 2-3 times faster on properly formatted URIs.

The benchmark:
```js
const bench = require('nanobench')
const fastDecode = require('fast-decode-uri-component')

const uri = [ 
  encodeURIComponent(' /?!#@=[](),\'"'),
  encodeURIComponent('много русских букв'),
  encodeURIComponent('acde=bdfd'),
  encodeURIComponent('這裡有些話'),
  encodeURIComponent('कुछ शब्द यहाँ'),
  encodeURIComponent('algunas palabras aquí'),
  encodeURIComponent('✌👀🎠🎡🍺')
]

function safeDecodeURIComponent(uri) {
  if (uri.indexOf('%') < 0) return uri
  try {
    return decodeURIComponent(uri)
  } catch (e) {
    return uri // or it can be null
  }
}

const uriLen = uri.length

bench('fast-decode-uri-component', b => {
  b.start()
  for (var round = 0; round < 100000; round++) {
    for (var i = 0; i < uriLen; i++) {
      fastDecode(uri[i])
    }
  }
  b.end()
})

bench('decodeURIComponent', b => {
  b.start()
  for (var round = 0; round < 100000; round++) {
    for (var i = 0; i < uriLen; i++) {
      safeDecodeURIComponent(uri[i])
    }
  }
  b.end()
})
```

As a result the standard function works 2.8 times faster:
```
# fast-decode-uri-component
ok ~512 ms (0 s + 511994600 ns)

# decodeURIComponent
ok ~182 ms (0 s + 182000400 ns)

all benchmarks completed
ok ~694 ms (0 s + 693995000 ns)
```

### PR DESCRIPTION
**Pull request by Eomm**
Closes #208 

The bench show how the `decodeURIComponent` vs `fast-decode-uri-component` has big differences based on the string that is being decoded.

I think `fast-decode-uri-component` is the best default for most use cases, but when the encoded character is more than 1 byte (UTF8) the `fast-decode-uri-component` performance drop.

This happens for no [Basic Latin](https://en.wikipedia.org/wiki/List_of_Unicode_characters#Basic_Latin) chars mainly.

It depends on the dev's implementation, of course, using alphanumeric IDs as path parameters do not decrease performance.

```

safeDecodeURIComponent(0) [' /?!#@=[](),\'"'] x 5,267,664 ops/sec ±0.23% (596 runs sampled)
fastDecode(0) [' /?!#@=[](),\'"'] x 3,136,283 ops/sec ±0.14% (596 runs sampled)
safeFastDecode(0) [' /?!#@=[](),\'"'] x 3,050,964 ops/sec ±0.17% (593 runs sampled)

safeDecodeURIComponent(1) [algunas palabras aqu%C3%AD] x 4,035,201 ops/sec ±0.14% (596 runs sampled)
fastDecode(1) [algunas palabras aqu%C3%AD] x 7,370,159 ops/sec ±0.14% (593 runs sampled)
safeFastDecode(1) [algunas palabras aqu%C3%AD] x 7,266,197 ops/sec ±0.24% (594 runs sampled)

safeDecodeURIComponent(2) [acde=bdfd] x 5,861,834 ops/sec ±0.16% (595 runs sampled)
fastDecode(2) [acde=bdfd] x 18,022,100 ops/sec ±0.20% (590 runs sampled)
safeFastDecode(2) [acde=bdfd] x 17,798,798 ops/sec ±0.24% (588 runs sampled)

safeDecodeURIComponent(3) [много русских букв] x 3,069,028 ops/sec ±0.17% (595 runs sampled)
fastDecode(3) [много русских букв] x 872,223 ops/sec ±0.18% (593 runs sampled)
safeFastDecode(3) [много русских букв] x 862,030 ops/sec ±0.29% (593 runs sampled)

safeDecodeURIComponent(4) [這裡有些話] x 5,252,010 ops/sec ±0.18% (594 runs sampled)
fastDecode(4) [這裡有些話] x 2,170,177 ops/sec ±0.15% (594 runs sampled)
safeFastDecode(4) [這裡有些話] x 2,153,915 ops/sec ±0.18% (594 runs sampled)

safeDecodeURIComponent(5) [कुछ शब्द यहाँ] x 3,520,488 ops/sec ±0.19% (595 runs sampled)
fastDecode(5) [कुछ शब्द यहाँ] x 842,579 ops/sec ±0.26% (595 runs sampled)
safeFastDecode(5) [कुछ शब्द यहाँ] x 837,686 ops/sec ±0.18% (595 runs sampled)

safeDecodeURIComponent(6) [✌👀🎠🎡🍺] x 4,212,286 ops/sec ±0.23% (593 runs sampled)
fastDecode(6) [✌👀🎠🎡🍺] x 1,650,773 ops/sec ±0.15% (595 runs sampled)
safeFastDecode(6) [✌👀🎠🎡🍺] x 1,691,189 ops/sec ±0.14% (596 runs sampled)

```

The nice-to-have would be optimizing the `fast-decode-uri-component` for this use case.

### CODE REVIEW

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by Eomm** [comment]
@delvedor could you rerun the failed jobs?

**Comment by Eomm** [comment]
mmmm the lint operation is failing on windows 😕 

image

**Comment by Eomm** [comment]
After #216 this CI should works

File: README.md [hunk 1]
```diff
@@ -323,6 +323,20 @@ Having a route with multiple parameters may affect negatively the performance, s
 
 **Note** that you must encode the parameters containing [reserved characters](https://www.rfc-editor.org/rfc/rfc3986#section-2.2).
 
+For [some use cases](https://github.com/delvedor/find-my-way/pull/211), you can customize the way the parameters are parsed by passing a custom decoding function.
```
Comment (mcollina) [inline README.md:326]: Please clarify what the use cases are instead of linking.

Comment (Eomm) [reply README.md]: done

File: README.md [hunk 2]
```diff
@@ -323,6 +323,20 @@ Having a route with multiple parameters may affect negatively the performance, s
...
+```js
+const router = require('find-my-way')({
+  decodeUriParameters: (stringToDecode) => {
+    // called when the request URL contains st least one special character: `# $ & + , / : ; = ? @`
+    return decodeURIComponent(stringToDecode)
+  }
+})
+```
+
+By default, this module relies on [fast-decode-uri-component`](https://www.npmjs.com/package/fast-decode-uri-component) to parse the encoded path parameters.
```
[the remark this answers is not in the document]
Comment (Eomm) [reply README.md:338]: To decode non basic Latin characters, I have added this detail above in the docs

**Review by mcollina — APPROVED** [review]
lgtm

**Review by delvedor — APPROVED** [review]
LGTM

**Pull request merged**